padl/transforms.py

```diff
@@ -2567,3 +2618,785 @@ def importdump(transform_or_module):
     t_copy = copy(transform_or_module)
     t_copy._pd_external_full_dump = False
     return t_copy
+
+
+class Node:
+    """Node wrapper for transforms
+
+    Nodes wrap a transform and can be connected in graph to form complex
+    graphs.
+    :param transform: Transform to wrap
+    """
+    _id = 0
+
+    def __init__(self, transform: Transform):
+        self.id = self._generate_id()  # keep?
+        self.transform = transform
+        self.pd_input_slice = self.transform._pd_input_slice
+        self.pd_output_slice = self.transform._pd_output_slice
+
+    @property
+    def name(self):
+        if self.transform.pd_name is None and isinstance(self.transform, Graph):
+            return self.transform._name
+        return self.transform.pd_name
+
+    @property
+    def name_id(self):
+        return self.name + ' ' + str(self.id)
+
+    @classmethod
+    def _generate_id(cls):
+        cls._id += 1
+        return cls._id
+
+    def __call__(self, args):  # should a node be callable?
+        """Call Method for Node"""
+        return self.call_node(args)
+
+    def copy(self):
+        _copy = type(self)(self.transform)
+        return _copy
+
+    def call_node(self, args):
+        return self.transform.pd_call_in_mode(args, mode='infer', ignore_grad=True)
+
+
+class Graph(Pipeline):
+    """Graph: New Pipeline
+
+    * Nodes listed: Should Contain Nodes to do the operation
+    * Edges are to be connected and only contained in a Graph
+    Example:
+    comp1 = A >> B >> C
+    comp2 = comp1 >> X >> Y
+    the connection from `comp1 >> X` should not leak in comp1
+    """
+
+    def __init__(self, transforms: Iterable[Transform], call_info: inspector.CallInfo = None,
+                 pd_name: Optional[str] = None, pd_group: bool = False):
+        super().__init__(transforms, call_info=call_info, pd_name=pd_name, pd_group=pd_group)
+
+        self.input_node = Node(identity - 'Input')  # potential overhead
+        self.output_node = Node(identity - 'Output')
+
+        self.networkx_graph = None  # ... later drop
+
+        self._pd_group = False
+        self.edges = defaultdict(dict)
+        self.parents = defaultdict(list)
+
+    def copy(self):
+        """Return copy of this graph"""
+        copy_graph = type(self)(self.transforms)
```

Comment:
Is it possible to use `copy_graph = copy(self)` here? If not you need to make sure to copy over all attributes inherited from `Transform` to the new copy as well. For example, 
```python
self._pd_call_info = call_info
self._pd_varname = {}
self._pd_name = pd_name
self._pd_device = 'cpu'
self._pd_traceback = traceback.extract_stack()
self._pd_external_full_dump = False
```
need to be all copied over as well, I believe.